File: 0044-wildcard-matching/solution.cpp
class Solution {
public:
    bool isMatch(string s, string p) 
    {
        const int n = s.size();
        int m = p.size();
        
        string new_p = "";
        bool is_first = true;
        int i = 0;
        while(i < m)
        {
            if(p[i] != '*')
            {
                new_p += p[i];
                i++;
            }
            while(i < m && p[i] == '*')
            {
                if(is_first){
                    new_p += p[i];
                    is_first = false;
                }
                i++;
            }
            is_first = true;
        }
        
        p = new_p;
        m = p.size();
        vector<bool> dp(m+1, false);
        dp[0] = true;
        if(m > 0 && p[0] == '*')
            dp[1] = true;
        
        for(i = 1; i < n+1; i++)
        {
            vector<bool> newDp(m+1, false);
            for(int j = 1; j < m+1; j++)
            {
                if(p[j-1] == '?' || s[i-1] == p[j-1])
                {
                    newDp[j] = dp[j-1];
                }
                else if(p[j-1] == '*')
                {
                    newDp[j] = newDp[j-1] || dp[j];
                }
            }
            dp = newDp;
        }
        
        return dp[m];

    }
};

File: 0044-wildcard-matching/solution.py
class Solution(object):
    
    def isMatch(self, s, p):
        """
        :type s: str
        :type p: str
        :rtype: bool
        """       
        n, m = len(s), len(p)
        
        new_p = []
        index = 0
        isFirst = True
        while index < m:
            if p[index] != '*':
                new_p.append(p[index])
                index += 1
            while index < m and p[index] == '*':
                if isFirst:
                    new_p.append("*")
                    isFirst = False
                index += 1
            isFirst = True
            
        p = "".join(new_p)
        m = len(p)

        # dp[i][j] checks if the substring s[0:i] matches the re p[0:j]
        dp = [False]*(m+1)
        if m and p[0] == '*':
            dp[1] = True
        dp[0] = True

        for i in xrange(1, n+1):
            new_dp = [False]*(m+1)
            for j in xrange(1, m+1):
                if p[j-1] == '?' or s[i-1] == p[j-1]:
                    new_dp[j] = dp[j-1]
                elif p[j-1] == '*':
                    new_dp[j] = new_dp[j-1] or dp[j]
            dp = new_dp

        return dp[m]
        
